### ISSUE TITLE
Variable is not generalized over

### ISSUE THREAD

**Issue by flupe** [agda/agda#6008]
I've come into another regression between current master and the latest release.
I'm struggling to find a minimal example, so here is a short file:

```agda
{-# OPTIONS --safe --without-K #-}
module generalization where

open import Agda.Primitive
open import Agda.Builtin.Sigma
open import Agda.Builtin.Unit
open import Agda.Builtin.Equality

private variable
  l : Level
  A : Set l
  a : A

data Tel (A : Set l) : Setω
levelOfTel : Tel A → Level
⟦_⟧tel     : (T : Tel A) → A → Set (levelOfTel T)

data Tel A where
  ε   : Tel A
  _⊢_ : (T : Tel A) {l : Level} (f : ∀ {a} → ⟦ T ⟧tel a → Set l) → Tel A

levelOfTel ε             = lzero
levelOfTel (_⊢_ T {l} f) = levelOfTel T ⊔ l

⟦ ε     ⟧tel a = ⊤
⟦ T ⊢ f ⟧tel a = Σ (⟦ T ⟧tel a) f

infix 4 _≡ⁿ_

_≡ⁿ_   : {T : Tel A} (xs ys : ⟦ T ⟧tel a) → Tel ⊤
reflⁿ  : {T : Tel A} {xs : ⟦ T ⟧tel a} → ⟦ xs ≡ⁿ xs ⟧tel tt
substⁿ : {T : Tel A} (f : ∀ {a} → ⟦ T ⟧tel a → Set l)
         {xs ys : ⟦ T ⟧tel a}
       → ⟦ xs ≡ⁿ ys ⟧tel tt
       → f xs → f ys
  
substⁿ-refl : {T : Tel A} (f : ∀ {a} → ⟦ T ⟧tel a → Set l)
              {xs : ⟦ T ⟧tel a} {x : f xs}
            → substⁿ f reflⁿ x ≡ x

_≡ⁿ_ = {!!}
reflⁿ = {!!}
substⁿ = {!!}
substⁿ-refl = {!!}
```
With the latest release, this typechecks fine and we end up with 4 holes and no other metavariables left to solve. The type of `substⁿ-refl` reduces to:

```agda
{a..1 : Level} {A : Set a..1} {l : Level} {a : A} {T : Tel A}
(f : {a = a₁ : A} → ⟦ T ⟧tel a₁ → Set l) {xs : ⟦ T ⟧tel a}
{x : f xs} →
substⁿ f reflⁿ x ≡ x
```

With current master though, the type of `substⁿ-refl` somehow does not generalize over the level `l` at which the (generalized) variable `A` is defined at, and introduces a metavariable instead. The type of `substⁿ-refl` reduces to:

```agda
{A : Set _A.l_140} {l : Level} {a : A} {T : Tel A}
(f : {a = a₁ : A} → ⟦ T ⟧tel a₁ → Set l) {xs : ⟦ T ⟧tel a}
{x : f xs} →
substⁿ f (_146 (f = f)) x ≡ x
```

where `_A.l_140` is the unexpectedly inserted metavariable. This leads to many unsolvable constraints, and my definitions don't go through.

Of course I can solve this issue by not using generalizable variables and explicitly quantify over everything, but this does look like a weird bug. As I said, I haven't been able to find shorter examples.

**Comment by flupe** [comment]
Bisecting seems to indicate that commit 982e18cdcbb is when this bug was introduced.

**Comment by andreasabel** [comment]
Here is a slight variation without holes:
```agda
{-# OPTIONS --without-K #-}

open import Agda.Primitive
open import Agda.Builtin.Sigma
open import Agda.Builtin.Unit
open import Agda.Builtin.Equality

postulate
  ANYω : ∀ {A : Setω} → A
  ANY  : ∀ {a} {A : Set a} → A

private variable
  l : Level
  A : Set l
  a : A

data Tel (A : Set l) : Setω
levelOfTel : Tel A → Level
⟦_⟧tel     : (T : Tel A) → A → Set (levelOfTel T)

data Tel A where
  ε   : Tel A
  _⊢_ : (T : Tel A) {l : Level} (f : ∀ {a} → ⟦ T ⟧tel a → Set l) → Tel A

levelOfTel ε             = lzero
levelOfTel (_⊢_ T {l} f) = levelOfTel T ⊔ l

⟦ ε     ⟧tel a = ⊤
⟦ T ⊢ f ⟧tel a = Σ (⟦ T ⟧tel a) f

infix 4 _≡ⁿ_

_≡ⁿ_   : {T : Tel A} (xs ys : ⟦ T ⟧tel a) → Tel ⊤
reflⁿ  : {T : Tel A} {xs : ⟦ T ⟧tel a} → ⟦ xs ≡ⁿ xs ⟧tel tt
substⁿ : {T : Tel A} (f : ∀ {a} → ⟦ T ⟧tel a → Set l)
         {xs ys : ⟦ T ⟧tel a}
       → ⟦ xs ≡ⁿ ys ⟧tel tt
       → f xs → f ys

substⁿ-refl : {T : Tel A} (f : ∀ {a} → ⟦ T ⟧tel a → Set l)
              {xs : ⟦ T ⟧tel a} {x : f xs}
            → substⁿ f reflⁿ x ≡ x

_≡ⁿ_        = ANYω
reflⁿ       = ANY
substⁿ      = ANYω
substⁿ-refl = ANY
```

**Comment by jespercockx** [comment]
This seems to be caused by the promotion of metavariables in the occurs checker that I introduced in https://github.com/agda/agda/commit/d1a9d7d5f5422701037ae48d780feeafd30a82c0, so it might have the same root cause as https://github.com/agda/agda/issues/5845

**Comment by andreasabel** [comment]
Github isn't good at understanding negations... ;-)
I wrote
> Unfortunately, this refinement does not fix https://github.com/agda/agda/issues/6008.

and it matched `fix #6008` and closed this issue.  Maybe I should have written:
> ... this refinement fixes not #6008.

**Comment by jespercockx** [comment]
Here is a minimized example:
```agda
postulate ANY  : ∀ {a} {A : Set a} → A

variable A : Set _

F : ∀ {l} (A : Set l) (a : A) → Set

postulate
  f   : ∀ {l} (A : Set l) (a : A) → F A a
  g   : (a : A) → F A a → Set
  g-f : (a : A) → g _ (f A a)

F A = ANY
```
Unsolved meta:
```
_A.1_30 : Agda.Primitive.Level
```
Output of `-v tc.generalize` (edit: with `--show-implicit`):
```current constraints:
  _28
    := λ genTel a →
         f {_A.1_20 (genTel = genTel)} (_A_21 (genTel = genTel)) a
    (blocked on problem 20)
  F {_A.1_20} _A_21 a =< F {_A.1_24} _A_25 _a_26
    (blocked on any(_A.1_20, _A.1_24, definition Issue6008.F),
     belongs to problem 20)
  constrainedMetas     =  _A.1_20, _A_21, _A.1_24, _A_25, _a_26, _28,
                          _29
  generalizable        =  _A_21
  generalizableOpen    =  _A_21
  openSortMetas        = 
  alsoGeneralize       = 
  reallyDontGeneralize =  _A.1_20, _A.1_24, _A_25, _a_26, _28, _29
we're generalizing over _A_21
```
It seems somehow the constraint `F _A_21 a =< F _A_25 _a_26` is preventing the generalization from happening.

**Comment by jespercockx** [comment]
The cause here seems indeed to be https://github.com/agda/agda/commit/982e18cdcbb2355d571bc05a8a98811fe6b48525, as applications of `F` used to be treated as neutrals in between its declaration and definition, but now they are treated as blocked instead. 

It seems that we are moving slowly from "not yet defined things are treated as postulates" to "not yet defined things are treated as metavariables", despite the discussion in https://github.com/agda/agda/issues/1113 pointing in the other direction. I am not necessarily against this, but this **is** a quite significant language change and not something we should take too lightly, since it already has caused a number of issues. @andreasabel @UlfNorell @nad @Saizan could you give your opinion on the matter?

**Comment by nad** [comment]
Can someone summarise the pros and cons of the two approaches? Should one ideally always use one approach, or does it make sense to use different approaches for different things?

**Comment by jespercockx** [comment]
Pro of treating not-yet-defined functions as postulates:
- Postulates are considered injective, so more metas can be solved.

Con of treating not-yet-defined functions as postulates:
- Postulates are rigid symbols, so you might get a hard type error instead of a constraint.

Pro of treating not-yet-defined functions as metavariables:
- The order of definitions in a mutual block does not influence whether Agda accepts it or not (see for example https://github.com/agda/agda/issues/2713#issuecomment-393585327 and https://github.com/agda/agda/issues/4618).

Cons of treating not-yet-defined functions as metavariables:
- It makes Agda accept mutual definitions where in order to typecheck, each definition relies on the computational behaviour of the other. In other words, there is no linear order in which one could typecheck the definitions, even with all implicit arguments given (see the example in https://github.com/agda/agda/issues/1113).
- It leads to unsolved constraints that might block variable generalization (as in this current issue).

**Comment by nad** [comment]
> Cons of treating not-yet-defined functions as metavariables:
> 
> * It makes Agda accept mutual definitions where in order to typecheck, each definition relies on the computational behaviour of the other. In other words, there is no linear order in which one could typecheck the definitions, even with all implicit arguments given (see the example in [In a mutual block, yet undefined things should be assumed to be metavariables, not postulates #1113](https://github.com/agda/agda/issues/1113)).

This sounds bad to me.

Is there a third approach that is better than the two listed above?

**Comment by jespercockx** [comment]
Not that I know of

**Comment by jespercockx** [comment]
I think we're too close to the release now to change the language, so I have reverted the fix of #5600 (and hence made Agda treat not yet defined functions as postulates again). If you have a better idea that avoids the problems I wrote above, please open a new issue for that.
